File: frontend/js/views/rent_roll.js
// Rent roll — multi-unit rental income summary, via /calc/rent-roll.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
const SEED = [
    { label: 'A', rent: 1500, occ: true, sqft: 800 },
    { label: 'B', rent: 1600, occ: true, sqft: 850 },
    { label: 'C', rent: 1400, occ: false, sqft: 750 },
    { label: 'D', rent: 1800, occ: true, sqft: 1000 },
];
function rowHtml(u) {
    return `<div class="mpb-row rr-row">
        <input type="text" class="rr-label" value="${esc(u.label || '')}">
        <input type="number" step="50" min="0" class="rr-rent" value="${u.rent}">
        <input type="number" step="50" min="0" class="rr-sqft" value="${u.sqft}">
        <label class="rr-occ"><input type="checkbox" class="rr-occ-cb" ${u.occ ? 'checked' : ''}></label>
        <button type="button" class="rr-del" data-i18n="view.rentroll.remove">Remove</button></div>`;
}
export async function renderRentRoll(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rentroll.h1.title">// RENT ROLL</span></h1>
        <p class="muted small" data-i18n="view.rentroll.hint.intro">A multi-unit rental income summary. From each unit's rent, occupancy, and square footage it computes scheduled vs actual rent, the vacancy loss, physical occupancy (units) and economic occupancy (rent), the annualized totals, and the average rent per square foot. Not investment advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.rentroll.h2.inputs">Units</h2>
        <form id="rr-form" class="inline-form">
            <label><span data-i18n="view.rentroll.label.property">Property</span><input type="text" name="property_label" value="Maple Court"></label>
        </form>
        <div class="mpb-head rr-head"><span data-i18n="view.rentroll.col.unit">Unit</span><span data-i18n="view.rentroll.col.rent">Rent ($)</span><span data-i18n="view.rentroll.col.sqft">Sq ft</span><span data-i18n="view.rentroll.col.occ">Occ</span><span></span></div>
        <div id="rr-rows">${SEED.map(rowHtml).join('')}</div>
        <button type="button" id="rr-add" class="secondary" data-i18n="view.rentroll.add">+ Add unit</button>
        </div><div id="rr-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#rr-form'); const rowsEl = mount.querySelector('#rr-rows');
    const gen = async () => {
        const units = [...rowsEl.querySelectorAll('.rr-row')].map((r) => ({ label: (r.querySelector('.rr-label').value || '').trim(), monthly_rent_usd: Number(r.querySelector('.rr-rent').value) || 0, occupied: r.querySelector('.rr-occ-cb').checked, sqft: Number(r.querySelector('.rr-sqft').value) || 0 })).filter((u) => u.label);
        const body = { property_label: (form.querySelector('[name="property_label"]').value || '').trim(), units };
        if (!units.length) { mount.querySelector('#rr-result').innerHTML = ''; return; }
        try { const d = await api.calcRentRoll(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.rentroll.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250);
    mount.querySelector('#rr-add').addEventListener('click', () => { rowsEl.insertAdjacentHTML('beforeend', rowHtml({ label: '', rent: 1500, occ: true, sqft: 800 })); applyUiI18n(rowsEl.lastElementChild); gen(); });
    rowsEl.addEventListener('click', (e) => { if (e.target.classList.contains('rr-del')) { e.target.closest('.rr-row').remove(); gen(); } });
    form.addEventListener('input', () => live()); rowsEl.addEventListener('input', () => live()); rowsEl.addEventListener('change', () => gen()); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#rr-result');
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card pos"><div class="label" data-i18n="view.rentroll.card.actual">Actual / month</div><div class="value">${money(d.actual_monthly_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.rentroll.card.sched">Scheduled / month</div><div class="value">${money(d.scheduled_monthly_usd)}</div></div>
        <div class="card neg"><div class="label" data-i18n="view.rentroll.card.vac">Vacancy loss</div><div class="value">${money(d.vacancy_loss_monthly_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.rentroll.card.phys">Physical occ.</div><div class="value">${pct(d.physical_occupancy_pct)} (${d.occupied_units}/${d.unit_count})</div></div>
        <div class="card"><div class="label" data-i18n="view.rentroll.card.econ">Economic occ.</div><div class="value">${pct(d.economic_occupancy_pct)}</div></div>
        <div class="card"><div class="label" data-i18n="view.rentroll.card.annual">Annual actual</div><div class="value">${money(d.annual_actual_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.rentroll.card.psf">Rent / sq ft</div><div class="value">${money(d.avg_rent_per_sqft_usd)}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/rent_roll.rs
//! Rent roll — a multi-unit rental income summary. For each unit it records the
//! monthly rent, whether it is occupied, and the square footage; from these it
//! computes scheduled rent (all units), actual rent (occupied only), the vacancy
//! loss, physical occupancy (occupied units ÷ total) and economic occupancy
//! (actual ÷ scheduled rent), the annualized totals, and the average rent per
//! square foot. Distinct from the single-property NOI module — this rolls up a
//! whole building. Pure compute, not investment advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Unit {
    pub label: String,
    pub monthly_rent_usd: f64,
    #[serde(default)]
    pub occupied: bool,
    #[serde(default)]
    pub sqft: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RentRollInput {
    pub property_label: String,
    pub units: Vec<Unit>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct RentRollReport {
    pub unit_count: usize,
    pub occupied_units: usize,
    pub scheduled_monthly_usd: f64,
    pub actual_monthly_usd: f64,
    pub vacancy_loss_monthly_usd: f64,
    /// Occupied units ÷ total units, percent.
    pub physical_occupancy_pct: f64,
    /// Actual rent ÷ scheduled rent, percent.
    pub economic_occupancy_pct: f64,
    pub annual_scheduled_usd: f64,
    pub annual_actual_usd: f64,
    /// Scheduled monthly rent ÷ total square footage (0 if no sqft given).
    pub avg_rent_per_sqft_usd: f64,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

pub fn generate(i: &RentRollInput) -> RentRollReport {
    let n = i.units.len();
    if n == 0 {
        return RentRollReport::default();
    }
    let scheduled: f64 = i.units.iter().map(|u| u.monthly_rent_usd).sum();
    let actual: f64 = i.units.iter().filter(|u| u.occupied).map(|u| u.monthly_rent_usd).sum();
    let occupied = i.units.iter().filter(|u| u.occupied).count();
    let total_sqft: f64 = i.units.iter().map(|u| u.sqft).sum();

    RentRollReport {
        unit_count: n,
        occupied_units: occupied,
        scheduled_monthly_usd: cents(scheduled),
        actual_monthly_usd: cents(actual),
        vacancy_loss_monthly_usd: cents(scheduled - actual),
        physical_occupancy_pct: round2(occupied as f64 / n as f64 * 100.0),
        economic_occupancy_pct: if scheduled > 0.0 { round2(actual / scheduled * 100.0) } else { 0.0 },
        annual_scheduled_usd: cents(scheduled * 12.0),
        annual_actual_usd: cents(actual * 12.0),
        avg_rent_per_sqft_usd: if total_sqft > 0.0 { round4(scheduled / total_sqft) } else { 0.0 },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn u(label: &str, rent: f64, occ: bool, sqft: f64) -> Unit {
        Unit { label: label.into(), monthly_rent_usd: rent, occupied: occ, sqft }
    }

    fn base() -> RentRollInput {
        RentRollInput {
            property_label: "Maple Court".into(),
            units: vec![
                u("A", 1500.0, true, 800.0),
                u("B", 1600.0, true, 850.0),
                u("C", 1400.0, false, 750.0),
                u("D", 1800.0, true, 1000.0),
            ],
        }
    }

    #[test]
    fn scheduled_actual_vacancy() {
        let d = generate(&base());
        assert_eq!(d.unit_count, 4);
        assert_eq!(d.occupied_units, 3);
        assert!(close(d.scheduled_monthly_usd, 6_300.0));
        assert!(close(d.actual_monthly_usd, 4_900.0));
        assert!(close(d.vacancy_loss_monthly_usd, 1_400.0));
    }

    #[test]
    fn physical_vs_economic_occupancy() {
        let d = generate(&base());
        assert!(close(d.physical_occupancy_pct, 75.0));
        assert!(close(d.economic_occupancy_pct, 77.78));
    }

    #[test]
    fn annualized_and_per_sqft() {
        let d = generate(&base());
        assert!(close(d.annual_scheduled_usd, 75_600.0));
        assert!(close(d.annual_actual_usd, 58_800.0));
        assert!(close(d.avg_rent_per_sqft_usd, 1.8529));
    }

    #[test]
    fn fully_occupied_equal_occupancy() {
        let mut inp = base();
        for unit in inp.units.iter_mut() {
            unit.occupied = true;
        }
        let d = generate(&inp);
        assert!(close(d.physical_occupancy_pct, 100.0));
        assert!(close(d.economic_occupancy_pct, 100.0));
        assert!(close(d.vacancy_loss_monthly_usd, 0.0));
    }

    #[test]
    fn empty_units_zero() {
        let d = generate(&RentRollInput { property_label: "X".into(), units: vec![] });
        assert_eq!(d.unit_count, 0);
        assert!(close(d.scheduled_monthly_usd, 0.0));
    }
}
